pkg/controller/usersignupcleanup/usersignup_cleanup_controller_test.go

```diff
@@ -143,6 +147,53 @@ func TestUserCleanup(t *testing.T) {
 		require.False(t, res.Requeue)
 	})
 
+	t.Run("test propagation policy", func(t *testing.T) {
+		userSignup := test2.NewUserSignup(
+			test2.CreatedBefore(threeYears),
+			test2.VerificationRequired(),
+		)
+
+		r, req, fakeClient := prepareReconcile(t, userSignup.Name, userSignup)
+
+		fakeClient.MockDelete = func(ctx context.Context, obj runtime.Object, opts ...client.DeleteOption) error {
+
+			for _, opt := range opts {
+				val := reflect.ValueOf(opt).Elem()
+				deletionPropagation := val.FieldByName("PropagationPolicy").Interface().(*v1.DeletionPropagation)
+				if *deletionPropagation == "Foreground" {
+					if obj, ok := obj.(*v1alpha1.UserSignup); ok {
+						util.AddFinalizer(obj, "foregroundDeletion")
+						if err := r.Client.Update(context.TODO(), obj); err != nil {
+							return err
+						}
+						return nil
+					}
+					return fmt.Errorf("object is not of type userSignup")
+				}
+			}
+			return fmt.Errorf("deletion propagation policy is not of type foreground")
+		}
+
+		_, err := r.Reconcile(req)
+		require.NoError(t, err)
+
+		// Confirm the UserSignup has finalizer set
+		key := test.NamespacedName(test.HostOperatorNs, userSignup.Name)
+		require.NoError(t, r.Client.Get(context.Background(), key, userSignup))
+		require.Equal(t, userSignup.Finalizers[0], "foregroundDeletion")
+
+		// now remove finalizer, reset MockDelete to nil and call reconcile
+		util.RemoveFinalizer(userSignup, "foregroundDeletion")
+		require.NoError(t, r.Client.Update(context.TODO(), userSignup))
+		fakeClient.MockDelete = nil
+
+		_, err = r.Reconcile(req)
+		require.NoError(t, err)
+
+		// now usersignup should be deleted
+		require.Error(t, r.Client.Get(context.Background(), key, userSignup))
```

Comment:
With this finalizer you are actually testing the mock function. Not the controller. Controller doesn't add any finalizer, at least explicitly. It's added by your mock function. The only thing we care here is if the correct propagation option was passed to the delete client. So I would simplify the test:
```suggestion
		deleted := false
		fakeClient.MockDelete = func(ctx context.Context, obj runtime.Object, opts ...client.DeleteOption) error {
			deleted = true
			require.Len(t, opts, 1)
			deleteOptions, ok := opts[0].(*client.DeleteOptions)
			require.True(t, ok)
			require.NotNil(t, deleteOptions)
			require.NotNil(t, deleteOptions.PropagationPolicy)
			assert.Equal(t, metav1.DeletePropagationForeground, *deleteOptions.PropagationPolicy)
			return nil
		}

		_, err := r.Reconcile(req)
		require.NoError(t, err)

		assert.True(t, deleted)
```

I actually mentioned it when saw that comment: https://github.com/codeready-toolchain/host-operator/pull/440/files#r636213278 days ago but forgot to press the submit button to submit the review :(

Reply:
@alexeykazakov oh my! You're right, it's more of a test for the MockDelete. I guess i got too caught up on the "The object's metadata.finalizers contains the value "foregroundDeletion" condition check". I will certainly replace with your commit. I hope I didn't misunderstand @xcoulon's comment aimed at checking finalizers. I believe the suggested change should indeed be enough to get the security wrt the propagation policy.